spring-kafka/src/main/java/org/springframework/kafka/core/reactive/ReactiveKafkaConsumerTemplate.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.core.reactive;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.consumer.OffsetAndTimestamp;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.kafka.receiver.KafkaReceiver;
+import reactor.kafka.receiver.ReceiverOptions;
+import reactor.kafka.receiver.ReceiverRecord;
+import reactor.kafka.sender.TransactionManager;
+
+/**
+ * Reactive kafka consumer operations implementation.
+ *
+ * @param <K> the key type.
+ * @param <V> the value type.
+ *
+ * @author Mark Norkin
+ */
+public class ReactiveKafkaConsumerTemplate<K, V> implements ReactiveKafkaConsumerOperations<K, V> {
```

Comment:
What benefit does this bring over Reactor Kafka?
Also I am a bit reluctant to make stateful operation in the template pattern...
However I see that we may consider it as a `JdbcTemplate`  which is really based on some `DataSource`, but I still don't see too much gain with this template over  standard Reactor Kafka `KafkaReceiver`...

Of course we can give it a shot and see how it makes sense with the target Spring Kafka users.

Reply:
Sorry, I didn't get the point about stateful operation in the template, can you elaborate please ?

As for usefulness of consumer template. At the moment it's mainly reactor-kafka receiver with a bit more friendlier api with a shortcuts, but I think that incorporating it into the framework also will give users familiar template abstraction.

Reply:
Yes, now see what you mean. Looking at the code of KafkaReceiver I see that it operates on a single kafka consumer, which is not thread safe and having the logic like
```
        if (!isActive.compareAndSet(false, true))
            throw new IllegalStateException("Multiple subscribers are not supported for KafkaReceiver flux");
``` 
probably gives the answer that it's shouldn't be used in that way, so the pattern as you prescribed would be to subscribe once to the flow and then layer other processing/sinking logic similar to what apache spark and other stream processing frameworks do.

Having template as a `factory` of receivers (if I understand you correctly) probably would be confusing, as each receiver bound to one kafka consumer, so creating a consumer each time would not work as expected for users.

As templates are based on KafkaReceiver/KafkaSender, I think it's better to align with reactor-kafka design for them.